### ISSUE TITLE
Unknown layer type error in tf_importer.cpp while running import->populateNet(Net)

### ISSUE THREAD

**Issue by mhaghighat** [opencv/opencv_contrib#1029]
##### System information (version)

- OpenCV => 3.2 (1aeeee0057faedbf6c693ce99ab36b12b07fc6d6)
- Operating System / Platform => Windows 64 Bit
- Compiler => Visual Studio 2015

##### Detailed description

I am trying to run the [tf_inception.cpp](https://github.com/opencv/opencv_contrib/blob/master/modules/dnn/samples/tf_inception.cpp) example. The compilation is successful, but it gives a run-time error. The error happens at [line 83](https://github.com/opencv/opencv_contrib/blob/master/modules/dnn/samples/tf_inception.cpp#L83) calling:
`importer->populateNet(net);`
 
`populateNet` goes through all layers in a for loop at [line 508](https://github.com/opencv/opencv_contrib/blob/master/modules/dnn/src/tensorflow/tf_importer.cpp#L508) of the [tf_importer.cpp](https://github.com/opencv/opencv_contrib/blob/master/modules/dnn/src/tensorflow/tf_importer.cpp). However, both name and type of the layer # 1 of the **tensorflow_inception_graph.pb** (when `li = 1`), is **DecodeJpeg**, which does not match any case in the loop and results in the error in [line 729](https://github.com/opencv/opencv_contrib/blob/master/modules/dnn/src/tensorflow/tf_importer.cpp#L729):

```
...
else
{
     printLayerAttr(layer);
     CV_Error_(Error::StsError, ("Unknown layer type %s in op %s", type.c_str(), name.c_str()));
}
```

**Comment by arrybn** [comment]
Hi, @mhaghighat! This is a problem due to there are some util layers in tensorflow graphs (like DecodeJpeg layer to load jpeg images). 
To launch tensorflow model in the dnn module you need firstly to remove all util layers from the model,
To do this, you can use this script: [freeze_graph.py](https://github.com/tensorflow/tensorflow/blob/master/tensorflow/python/tools/freeze_graph.py) (and you need to install python and tensorflow). This script removes all things which are irrelevant to inference and saves cleared model, so you need to launch this script only once.
The script has next launch parameters which you should specify:
--input_graph=/hdd/tf_models/alexnet_full/conv_graph.pb  (input .pb file with graph description)
--input_checkpoint=/hdd/tf_models/alexnet_full/conv_checkpoint.ckpt   (input .ckpt ckeckpoint file with weights)
--output_graph=/hdd/tf_models/alexnet_full/conv_graph_frozen.pb  (path to output cleared graph)
--output_node_names=prob  (name of the layer with output you're interested in)
--input_binary  (specify input files format)
Please let me know If you have any problems with this script.
I'll consider adding feature of removing util layers when loading tf models in dnn module

**Comment by arrybn** [comment]
My post above is for case when you've trained your own model tensorflow and have .pb and .ckpt files.
Regarding to your case, I've checked once more and model from here: https://storage.googleapis.com/download.tensorflow.org/models/inception5h.zip works as is in dnn module. Please check that you're using this model.
It's possible that some layers (specific) are missed in current release of dnn module. DecodeJpeg layer is util layer, it can be replaced by Identity layer when parsing and I'll try to do it soon. I'll let you know if I suggest solution for this issue.

**Comment by mhaghighat** [comment]
Hi @arrybn,
Thank you for your reply. I am using the model in your provided link (the '_tensorflow_inception_graph.pb_' frozen model), and I still have that issue. 
How do you suggest that I should replace the util layers, like DecodeJpeg, with Identity? Won't it affect the result?

**Comment by arrybn** [comment]
Hi, @mhaghighat. I've checked once more time: downloaded the archive from here: https://storage.googleapis.com/download.tensorflow.org/models/inception5h.zip, unzipped it and used tensorflow_inception_graph.pb model file. And c++ sample tf_inception.cpp works correctly. Also I've checked python interfaces - also no problems, model has been loaded and output has been computed. Please check launch parameters for the sample, maybe it uses previously downloaded model from other source.
As I've found, there are 2 types of inception model distributed by Google - one with DecodeJpeg and other layers like subtract mean, normalize and so on (call them preprocessed layers because they don't have learnable parameters and do conversion of some output to form which appropriate for the model); and another - without all preprocessing layers.
If you remove preprocessing layers and do all removed operations by yourself all will stay correct.
In case of our sample tf_inception.cpp we manually decode  images from jpeg by cv::imread() and can do any preprocessing.
To sum up, please double check paths (maybe remove all inception model you've already downloaded, and download new from the provided link) because all works fine for me

**Comment by arrybn** [comment]
Hello, @mhaghighat. Are you still experiencing described issue?

**Comment by chvatma2** [comment]
Hello, 
I have encountered the same issue as described here. We are using the inception model, retrained on our images as shown in this tutorial: https://www.tensorflow.org/tutorials/image_retraining. The suggested freeze_graph.py script did not solve the problem. It would be useful if the dnn module could handle models created this way as the official Tensorflow tutorials will surely be an entry point for many programmers.

**Comment by chrisrn** [comment]
@arrybn  I have 2 questions:

1) The freeze_graph script removes the processing nodes or we need to do it manually?
2) I tried to load the inception5h model from your link and the setBlob function cannot recognize the input name (Requested blob 'input' not found). I saw that the input node from this model is called 'input' and the output node is called 'output'. You fed different values to i_blob and o_blob?

**Comment by arrybn** [comment]
1. As far as I know, the freeze_graph script save all Variable nodes as Const and removes some other utility  nodes. But it doesn't change preprocessing layers such mean-variance normalization, DecodeJpeg and other nodes. You should do it manually in tensorflow.
2. There is an [example](https://github.com/opencv/opencv_contrib/blob/master/modules/dnn/samples/tf_inception.cpp) of using inception5h in tensorflow in dnn. Use ".input" as name for input layer and "softmax2" as output name (see sample)

**Comment by chrisrn** [comment]
Thank you it worked, I could just use the default values! Ok so I will try to remove the preprocessing layers and I will inform you if it works for my model.

**Comment by chrisrn** [comment]
@arrybn  I don't think that the freeze_graph script removes layers. In my case I used a specific way of reading tensorflow images from TFR format using queues. The 'FIFOQueueV2' layer is not removed after running freeze_graph. And I think it's not a good solution to change the way of reading and restart training because when you have to do with big data, tensorflow proposes this way of reading.

**Comment by chrisrn** [comment]
I can do the forward pass but the getBlob function reports error that cannot find the output. And it's pretty clear that the output exists in the graph. Any suggestions?

**Comment by arrybn** [comment]
@chrisrn I think output blob has name which differs from "output". You can get all blobs names by invoking Net::getLayerNames(), and it's highly likely that the last name from the returned vector is for output blob

**Comment by yxl10** [comment]
@arrybn 
Hi, I loaded the inception5h model from your link and tf_inception.cpp example works. However it gives the same class name for every image. The best class is always #0.
**Best class: #0 'dummy'
Probability: -4.31602e+10%**

 Could you help me?

**Comment by chrisrn** [comment]
@arrybn You can run it if you export the graph from a simplified python predictor without the preprocessing layers. But I see some differences when I predict using the tf_inception script (for my model) and I cannot understand why. The steps are the following:

1) Export .pbtxt from python predictor.
2) Feed the weights using freeze_graph script from tensorflow and export .pb file.
3) Predict by feeding the .pb file to the tf_inception script.

**Comment by arrybn** [comment]
@chrisrn could you share with me all necessary files: .pbtxt, .pb ? And what is python predictor?

**Comment by chrisrn** [comment]
By "python predictor" I mean that you don't need to export the graph during training. You can create a tensorflow script only for prediction without the preprocessing layers. Tensorflow has to do with more layers during training that's why I mentioned that.

**Comment by arrybn** [comment]
To makes things clear:
1. You have right_graph_gray1.txt and right_graph_gray1.cpkt files after training.
2. You use freeze_graph.py to turn them into .pb unite file (call this file deploy.pb)
3. Output of tensorflow differs from output of dnn then you're launching deploy.pb with some input.
Is all correct?
If yes, first check that inputs are the same. If image is in jpeg format, it's highly recommended to use png or other non-compressing formats. The reason is different frameworks and libs use different implementations of jpeg decoding

**Comment by chrisrn** [comment]
Everything right. The only difference is at step 1 where I am exporting the pbtxt file from another script in which the preprocessing layers do not exist. Inputs are the same that's not the problem. I am handling .avi video file using opencv on both python and c++ prediction.

**Comment by arrybn** [comment]
But tensorflow accepts NHWC-ordered blobs, and dnn's blobFromImage() returns NCHW

**Comment by chrisrn** [comment]
So you propose the conversion to NHCW?

**Comment by arrybn** [comment]
I just want to put your attention to this fact and to my mind it can be the issue. Could you send me your .pb file which you're loading to dnn?

**Comment by arrybn** [comment]
Suppose you have frame from OpenCV's VideoCapture. This is a 3-channels BGR numpy.ndarray with HWC dimensions order. Tensorflow accepts 4-dimensional blobs and you can add one dimension to the input image (to make it 4-dimensional NHWC) and pass it to tensorflow. But dnn works with NCHW. So you need to reorder dimansions. You can do it by numpy.transpose() function

**Comment by chrisrn** [comment]
[deploy.txt](https://github.com/opencv/opencv_contrib/files/1088339/deploy.txt)

.pb format is not supported that's why it's txt. You can convert it back to its normal .pb format.

**Comment by chrisrn** [comment]
I think that if I reorder the dimensions of the image that enters the dnn::Blob::fromImages function it can work.

**Comment by arrybn** [comment]
It seems that you're using old version of the dnn. Consider updating to the latest master, because it becomes faster, with low memory consumption and without some critical bugs

**Comment by chrisrn** [comment]
I use opencv-3.2 so I have the equivalent opencv-contrib-3.2 modules.

**Comment by arrybn** [comment]
Now in the dnn we remove Blob class, that's why I thought you're using previous version

**Comment by chrisrn** [comment]
So I need to build again opencv with the modules of master branch?

**Comment by arrybn** [comment]
Yes, you need to ckeckout the latest opencv and opencv_contrib master braches and after build

**Comment by chrisrn** [comment]
I did that but right now I am getting segmentation fault on net.forward command. Is this because I need to convert the image to NCHW before inserting the network?

**Comment by arrybn** [comment]
Probably yes. It would be perfect if you share your python script and tensorflow model so I can reproduce this bug on my workstation. The code can be as compact as possible and you can post it here. Without the code it will be a random guessing

**Comment by chrisrn** [comment]
It's difficult to share it because I am handling specific avi files and I export the graph into pb file. But I uploaded this file yesterday so you can reproduce the error only from the tf_inception script.

**Comment by arrybn** [comment]
You can rewrite it to use image instead of video. Try to make the simpliest sample, which reproduces the problem. It will help to solve your problem much faster

**Comment by chrisrn** [comment]
You cannot run it because it's a script that predicts and you need my checkpoint file which is too big. Anyway, after the blobFromImage function I noticed that the inputBlob has shape (1, 1, 32, 32) so I reshaped it again to (1, 32, 32, 1) just like the blobFromImage function to enter the tensorflow model correctly, but I am getting memory corruption error on copyTo function!

**Comment by arrybn** [comment]
dnn accepts  (1, 1, 32, 32) blob, but for tensorflow it should be (1, 32, 32, 1)

**Comment by chrisrn** [comment]
Yes sure. But if you check the pbtxt file the input has shape (1, 32, 32, 1). That's wrong?

**Comment by arrybn** [comment]
Could you post console output? Does it contain some layer names, types and another useful information?

**Comment by chrisrn** [comment]
Nop. It compiles well but when running the exe it produces Segmentation fault (core dumped). I know that it's difficult to understand but thank you very much for your help. The initial problem was that I did not have good predicitions by running the tf_inception script for my model and when I changed to the latest version of opencv-contrib I am getting the seg fault.

**Comment by arrybn** [comment]
Could you modify the dnn's code a little bit? If so, insert this line:
`std::cout << ld.name << " " << ld.type << std::endl;`
in the beginning of void forwardLayer(LayerData &ld) function ( modules/dnn/src/dnn.cpp, line 1016)
Save and rebuild
It prints you name and type of layer. The last printed entry will be for layer with segfault

**Comment by chrisrn** [comment]
Thank you for this tip. The console output is:
```
_input __NetInputLayer__
CifarNet/conv1/convolution Convolution
```
It is complaining for the 1st layer of the network, after the weights initialization.

**Comment by arrybn** [comment]
I think the only way to solve the problem is to debug. You should either do it by yourself or send me a file with the model. Also you can save the model in binary format instead of textual. It reduces size of a file a lot. Also you can initialize the model with random weights, if you don't want to share trained ones. Or you can remove all layers except this convolution

**Comment by chrisrn** [comment]
I uploaded the `right_graph_gray1.txt` file which contains the graph. I cannot upload the binary file here I don't know why. But I think you can convert this file into .pb file by the following snippet:
```
import tensorflow as tf
from tensorflow.python.platform import gfile
from google.protobuf import text_format

filename = '/path/to/right_graph_gray1.pbtxt'
with gfile.FastGFile(filename, 'r') as f:
    graph_def = tf.GraphDef()
    text_format.Merge(f.read(), graph_def)
    tf.import_graph_def(graph_def, name='')
    tf.train.write_graph(graph_def, '', 'deploy.pb', as_text=False)
```

**Comment by chrisrn** [comment]
I also tested my tf_inception script on the original tensorflow model and the segmentation fault still exists with console output:
```
_input __NetInputLayer__
conv2d0_pre_relu/conv Convolution
```
I do not predict only one image, but I feed images from a video to the network, so probably it has to do with overloading somehow?  Anyway I will try debugging.

**Comment by chrisrn** [comment]
I just moved to the newest version of opencv and I don't have the seg fault. But the predictions are still wrong for my model. I think you have to explain more precisely in the documentation how can we load our own models, because a lot of mistakes are happening. Thanks for the whole help in this issue!

**Comment by dkurt** [comment]
@chrisrn, I've tested the last one of your networks.
```
output of TensorFlow: [[  9.99834776e-01   1.65201331e-04]]
output of DNN:        [[  9.99834776e-01   1.65201651e-04]]
```

Code:
```python
import numpy as np
import tensorflow as tf
import cv2 as cv

# Read frozen model.
with tf.gfile.FastGFile('deploy.txt') as f:
    graph_def = tf.GraphDef()
    graph_def.ParseFromString(f.read())
    for node in graph_def.node:
        print node.op, node.name

with tf.Session() as sess:
    # Restore session
    sess.graph.as_default()
    tf.import_graph_def(graph_def, name='')

    # Generate input
    np.random.seed(2701)
    inp = np.random.standard_normal([1, 32, 32, 1]).astype(np.float32)

    # Receive output
    outTensor = sess.graph.get_tensor_by_name('softmax2:0')
    out = sess.run(outTensor, feed_dict={'input:0': inp})
    print out

def NHWCtoNCHW(data):
    dataSwap = data.swapaxes(1, 3)
    dataSwap = dataSwap.swapaxes(2, 3)
    return dataSwap

# Load network
net = cv.dnn.readNetFromTensorflow('deploy.txt')
# Set input in appropriate data format
net.setInput(NHWCtoNCHW(inp))
# Receive output
cvOut = net.forward()

print cvOut
```

**Comment by chrisrn** [comment]
Thank you very much I already solved that. The problem did not have to do with the preprocessing, so finally the dnn module works fine! Thanks for your interest!

**Comment by dkurt** [comment]
The problems seems to me resolved.
To highlight steps how to use models trained in TensorFlow:
* Be sure that graph has at least one variable (trainable parameter).
* Save session.
```python
saver = tf.train.Saver()
saver.save(sess, 'model.ckpt')
```
You'll receive something like
```
checkpoint
model.ckpt.data-00000-of-00001
model.ckpt.index
model.ckpt.meta
```

* Save graph in textual format.
```python
tf.train.write_graph(sess.graph.as_graph_def(), "", "graph.pbtxt")
```

* Freeze graph. Because text representation doesn't contain weights, we need to merge "graph.pbtxt" with checkpoint.
```bash
python freeze_graph.py
  --input_graph graph.pbtxt
  --input_checkpoint model.ckpt
  --output_graph frozen_graph.pb
  --output_node_names "name_of_output_op_here"
```

* Optimize for inference. This tool removes all training nodes from graph.
```bash
python optimize_for_inference.py
  --input frozen_graph.pb
  --output final_network.pb
  --frozen_graph True
  --input_names "name_of_input_op_here"
  --output_names "name_of_output_op_here"
```

`final_network.pb` might be successfully imported in DNN.

See tools @ https://github.com/tensorflow/tensorflow/tree/master/tensorflow/python/tools.

**Comment by littlesun96** [comment]
Hello, @dkurt !
I'm working with model with LSTM, trained in Keras using Tensorflow backend, and I also have the issue "OpenCV Error: Unspecified error (Unknown layer type StridedSlice in op lstm_1/strided_slice) 
I used scripts freeze_graph.py and optimize_for_inference.py, as written before, but nothing had changed. 
Maybe you can help?

**Comment by dkurt** [comment]
@littlesun96, Hi! We can help you faster with some piece of code. Tensorflow usually has several implementations of layers/nodes (i.e. at `tf.nn`, `tf.layers`, `tf.contrib.layers`, `tf.contrib.keras.layers`). Please show the way you've added LSTM node.

**Comment by littlesun96** [comment]
`from keras.layers import Input, LSTM, Dense
from keras.models import Model

main_input = Input(shape=(seq_len,3), name='main_input')
lstm_out = LSTM(20, return_sequences=False, input_shape=(seq_len,3))(main_input)
label_out = Dense(1, activation = 'sigmoid', name = 'label_output')(lstm_out)
model = Model(inputs = [main_input], outputs = [label_out])`

**Comment by dkurt** [comment]
@littlesun96, We've started working on it. I think, you'd better create a single issue and assign it to me.

**Comment by littlesun96** [comment]
@dkurt, Okay, thank you!

**Comment by BAILOOL** [comment]
I have tried to import my TF(v1.2) model to opencv 3.3 (ubuntu 16.04) after running `freeze_graph.py` and `optimize_for_inference.py`, but faced the problem with the `tf.nn.conv2d_transpose()`. Specifically, I get the following error:

> error: (-2) Unknown layer type Conv2DBackpropInput in op objectness_uplayer-0/deconv_layer in function populateNet

Supposedly, Deconvolution layer (cv::dnn::DeconvolutionLayer) is supported as mentioned in documentation, and FCN [sample](https://github.com/opencv/opencv/blob/master/samples/dnn/fcn_semsegm.cpp).  Any idea how to solve this problem?

**Comment by dkurt** [comment]
@BAILOOL, `Conv2DBackpropInput` was merged recently with https://github.com/opencv/opencv/pull/9349 with a test case https://github.com/opencv/opencv_extra/blob/master/testdata/dnn/tensorflow/generate_tf_models.py#L111. May I ask you to update source code and try again?

**Comment by BAILOOL** [comment]
@dkurt Indeed up-to-date version of opencv does not have the issue with `Conv2DBackpropInput`. Unfortunately, using `tf.slice` similarly results  in

> error: (-2) Unknown layer type Slice in op

**Comment by dkurt** [comment]
@BAILOOL, could you show a minimal graph which use `tf.slise`? May be we could find an alternative solution.

**Comment by Devy001** [comment]
@dkurt  I have ran the code freeze_graph.py and optimize_for_inference.py,and there is a final_network.pb generated, when I imported the final_network.pb in DNN with opencv,there is something wrong like this:
batch/fifo_queue:FIFOQueueV2
capacity:[ ]
component_types:[ ]
_output_shapes:[ ]
container:[ ]
shapes:[ ]
shared_name:[ ]
OpenCV Error: Unspecified error (Unknown layer type FIFOQueueV2 in op batch/fifo_queue) in populateNet, file /home/mi/opencv-3.4.0/modules/dnn/src/tensorflow/tf_importer.cpp, line 1487
terminate called after throwing an instance of 'cv::Exception'
  what():  /home/mi/opencv-3.4.0/modules/dnn/src/tensorflow/tf_importer.cpp:1487: error: (-2) Unknown layer type FIFOQueueV2 in op batch/fifo_queue in function populateNet
I do not know how to solve it, could you tell me ? thank you.

**Comment by dkurt** [comment]
@Devy001, please don't use closed issues. It's difficult to track. Ask a question at http://answers.opencv.org using appropriate tags. Attach a model or provide a minimal working example. We can't fix your model remotely.

**Comment by Devy001** [comment]
@dkurt thank you, I will submit the question on http://answers.opencv.org and upload a model later.

**Comment by Devy001** [comment]
@dkurt I am sorry to bother you again, I have submitted the question on the link that http://answers.opencv.org/question/183599/unknown-layer-type-fifoqueuev2-in-op-batchfifo_queue-in-function-populatenet/
thank you.

**Comment by Swordancer** [comment]
I have similar problem...

I have model generated with this tutorial for my own classes: https://www.tensorflow.org/tutorials/image_retraining

I'm getting an error:

> Unknown layer type Sub in op module_apply_default/hub_input/Sub

So I was trying to use freeze_graph.py

> py C:\tensorflow-master\tensorflow\python\tools\freeze_graph.py --input_graph=output_graph.pb --input_checkpoint=C:/tmp/_retrain_checkpoint --output_graph=/tmp/frozen_graph.pb --output_node_names=softmax --input_binary=true

So error again:

> Names_to_saveables must be a dict mapping string names to Tensors/Variables. Not a variable: Tensor("final_retrain_ops/biases/final_biases:0", shape=(6,), dtype=float32)

Then maybe optimize_for_inference.py
Guess what...

Multiple error xD

> WARNING:tensorflow:Didn't find expected Conv2D input to 'module_apply_default/MobilenetV2/expanded_conv_16/depthwise/BatchNorm/FusedBatchNorm'

Im trying to solve these problems from week. Can anyone help with that me?
